aten/src/ATen/core/type.cpp

```diff
@@ -200,14 +188,42 @@ c10::optional<TypePtr> unifyTypes(const TypePtr& t1, const TypePtr& t2) {
       }
     }
     return static_cast<TypePtr>(TupleType::create(elements));
-  } else if (t1->cast<DictType>() && t2->cast<DictType>()) {
+  }
+
+  // Handle direct subtyping relations
+  if (t1->isSubtypeOf(t2)) {
+    return t2;
+  } else if (t2->isSubtypeOf(t1)) {
+    return t1;
+  }
+
+  // Now, check subtyping relations again with Unshaped Types,
+  // to handle unification of container types which might contain two different
+  // specialized tensors (ListType / FutureType)
+  t1 = unshapedType(t1);
+  t2 = unshapedType(t2);
+
+  if (t1->isSubtypeOf(t2)) {
+    return t2;
+  } else if (t2->isSubtypeOf(t1)) {
+    return t1;
+  }
+
+  // because we have runtime specializations of lists, e.g. int[] = std::vector<int64_t>
+  // int?[] = std::vector<IValue>  we don't allow type coercion,
+  // since t1 & t2 may have different runtime representations.
+  // if we did not have runtime specialization, we could try to unify the List element type.
+  // as is, List unification is covered by direct subtyping relation check above
```

Comment:
FWIW another way of saying this is because of specialized representation of Lists, the usual rule 
```
unify(Container[t1, ...], Container[t2, ...]) => Container[unify(t1, t2), ...]
``` 
doesn't hold for `List`.